### PR TITLE
internal/sgx: do not share quote between multiple issuers

### PR DESCRIPTION
**Pull request by avalluri**
CTK destroys the quote public key and the wrapped key after successful unwrap. Hence the same quote cannot be used for unwrapping the other keys. So, we have to use singer specific quote.

### CODE REVIEW

**Comment by avalluri** [comment]
The PR is rebased on #33

**Comment by ipuustin** [comment]
Needs rebase against the new changes.

File: internal/sgx/sgx.go [hunk 1]
```diff
@@ -249,6 +269,11 @@ func (ctx *SgxContext) RemoveSigner(name string) error {
 
 	s := ctx.signers.Get(name)
 	if s == nil {
+		if quoteInfo, ok := ctx.quotes[name]; ok {
+			ctx.p11Ctx.DestroyObject(ctx.p11Session, quoteInfo.prvKeyHandle)
```
Comment (ipuustin) [inline internal/sgx/sgx.go:273]: Is there a need to check if the handles exist?

File: internal/sgx/sgx.go [hunk 2]
```diff
@@ -357,6 +382,11 @@ func (ctx *SgxContext) provisionKey(signerName string, wrappedSWK []byte, wrappe
 	ctx.ctxLock.Lock()
 	defer ctx.ctxLock.Unlock()
 
+	quoteInfo := ctx.quotes[signerName]
```
Comment (ipuustin) [inline internal/sgx/sgx.go:385]: Maybe follow the same `quoteInfo, ok :=` style as above?

File: internal/sgx/sgx.go [hunk 3]
```diff
@@ -608,14 +611,46 @@ func (ctx *SgxContext) generateQuote(pubKey pkcs11.ObjectHandle) ([]byte, error)
...
+	if bigExponent.BitLen() > 32 || bigExponent.Sign() < 1 {
+		return nil, fmt.Errorf("malformed quote public key")
+	}
+	if bigExponent.Uint64() > uint64(math.MaxInt) {
+		return nil, fmt.Errorf("malformed quote public key: possible data loss in exponent value")
 	}
-	return quotePubKey[offset:], nil
+	exponent := int(bigExponent.Uint64())
+	var modulus = new(big.Int)
+	modulus.SetBytes(pubkey[paramsSize+exponentLen:])
```
Comment (ipuustin) [inline internal/sgx/sgx.go:649]: Is there still a need to check the length here? To prevent out-of-bounds error.

**Review by ipuustin — COMMENTED** [review]
Just questions, code looks fine.

File: internal/sgx/sgx.go (hunk 1 above)
Comment (avalluri) [reply internal/sgx/sgx.go:273]: I donät think `quoteInfo` with invalid handles gets stored in the map.

File: internal/sgx/sgx.go (hunk 2 above)
Comment (avalluri) [reply internal/sgx/sgx.go]: Done.

File: internal/sgx/sgx.go (hunk 3 above)
Comment (avalluri) [reply internal/sgx/sgx.go]: Yes, the check is needed. I added it now.

**Review by ipuustin — APPROVED** [review]
Looks good to me now!

**Pull request merged**